### PR TITLE
python312Packages.kaleido: init at 0.2.1

### PR DESCRIPTION
**Pull request by Pandapip1**
## Description of changes

I'm sorry, I literally couldn't figure out how to do a source build. If someone would like to figure it out for themselves, uh, good luck!

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/kaleido/default.nix [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+      };
+      aarch64-darwin = fetchurl {
+        url = "https://files.pythonhosted.org/packages/py2.py3/k/kaleido/kaleido-0.2.1-py2.py3-none-macosx_11_0_arm64.whl";
+        hash = lib.fakeHash;
+      };
+    }
+    ."${system}" or (throw "Unsupported system for ${pname}: ${system}");
+
+  nativeBuildInputs = [ autoPatchelfHook ];
+  buildInputs = [
```
Comment (natsukium) [inline pkgs/development/python-modules/kaleido/default.nix:45]:
```suggestion
  buildInputs = [
    bash
```
It needs bash to modify shebang.
<summary>error log</summary><details>

```
❯ python
Python 3.12.5 (main, Aug  6 2024, 19:08:49) [GCC 13.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import plotly.express as px
>>> fig = px.scatter(px.data.iris(), x="sepal_length", y="sepal_width", color="species")
>>> fig.write_image("figure.png", engine="kaleido")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/nix/store/rjdw9s3lpvk6sfpncrj89xn7pa9q8mwm-python3-3.12.5-env/lib/python3.12/site-packages/plotly/basedatatypes.py", line 3835, in write_image
    return pio.write_image(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/rjdw9s3lpvk6sfpncrj89xn7pa9q8mwm-python3-3.12.5-env/lib/python3.12/site-packages/plotly/io/_kaleido.py", line 266, in write_image
    img_data = to_image(
               ^^^^^^^^^
  File "/nix/store/rjdw9s3lpvk6sfpncrj89xn7pa9q8mwm-python3-3.12.5-env/lib/python3.12/site-packages/plotly/io/_kaleido.py", line 143, in to_image
    img_bytes = scope.transform(
                ^^^^^^^^^^^^^^^^
  File "/nix/store/rjdw9s3lpvk6sfpncrj89xn7pa9q8mwm-python3-3.12.5-env/lib/python3.12/site-packages/kaleido/scopes/plotly.py", line 153, in transform
    response = self._perform_transform(
               ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/rjdw9s3lpvk6sfpncrj89xn7pa9q8mwm-python3-3.12.5-env/lib/python3.12/site-packages/kaleido/scopes/base.py", line 293, in _perform_transform
    self._ensure_kaleido()
  File "/nix/store/rjdw9s3lpvk6sfpncrj89xn7pa9q8mwm-python3-3.12.5-env/lib/python3.12/site-packages/kaleido/scopes/base.py", line 177, in _ensure_kaleido
    self._proc = subprocess.Popen(
                 ^^^^^^^^^^^^^^^^^
  File "/nix/store/pgb120fb7srbh418v4i2a70aq1w9dawd-python3-3.12.5/lib/python3.12/subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/nix/store/pgb120fb7srbh418v4i2a70aq1w9dawd-python3-3.12.5/lib/python3.12/subprocess.py", line 1955, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '/nix/store/rjdw9s3lpvk6sfpncrj89xn7pa9q8mwm-python3-3.12.5-env/lib/python3.12/site-packages/kaleido/executable/kaleido'
```
</details>

I suggest running the examples in `passthru.tests` to see if the figures are generated.
https://github.com/plotly/Kaleido?tab=readme-ov-file#use-kaleido-to-export-plotlypy-figures-as-static-images

File: pkgs/development/python-modules/kaleido/default.nix [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  libGL,
+  nspr,
+  nss,
+  sqlite,
+  system ? builtins.currentSystem,
+}:
+
+buildPythonPackage rec {
+  pname = "kaleido";
+  version = "0.2.1.post1";
```
Comment (natsukium) [inline pkgs/development/python-modules/kaleido/default.nix:16]:
```suggestion
  version = "0.2.1";
```
This version just adds support for arm32.
https://github.com/plotly/Kaleido/releases

File: pkgs/development/python-modules/kaleido/default.nix [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  format = "wheel";
+
+  src =
+    let
+      pypiPackageName = "kaleido";
+    in
+    {
+      # This library is so cursed that I have to use fetchurl instead of fetchPypi. I am not happy.
+      x86_64-linux = fetchurl {
+        url = "https://files.pythonhosted.org/packages/py2.py3/k/kaleido/kaleido-0.2.1-py2.py3-none-manylinux1_x86_64.whl";
```
Comment (natsukium) [inline pkgs/development/python-modules/kaleido/default.nix:26]:
```suggestion
        url = "https://files.pythonhosted.org/packages/py2.py3/k/kaleido/kaleido-${version}-py2.py3-none-manylinux1_x86_64.whl";
```

File: pkgs/development/python-modules/kaleido/default.nix [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+buildPythonPackage rec {
+  pname = "kaleido";
+  version = "0.2.1.post1";
+  format = "wheel";
+
+  src =
+    let
+      pypiPackageName = "kaleido";
+    in
```
Comment (natsukium) [inline pkgs/development/python-modules/kaleido/default.nix:22]: unused

File: pkgs/development/python-modules/kaleido/default.nix [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      x86_64-darwin = fetchurl {
+        url = "https://files.pythonhosted.org/packages/py2.py3/k/kaleido/kaleido-0.2.1-py2.py3-none-macosx_10_11_x86_64.whl";
+        hash = lib.fakeHash;
+      };
+      aarch64-darwin = fetchurl {
+        url = "https://files.pythonhosted.org/packages/py2.py3/k/kaleido/kaleido-0.2.1-py2.py3-none-macosx_11_0_arm64.whl";
+        hash = lib.fakeHash;
+      };
+    }
+    ."${system}" or (throw "Unsupported system for ${pname}: ${system}");
```
Comment (natsukium) [inline pkgs/development/python-modules/kaleido/default.nix:42]:
```suggestion
    ."${stdenv.hostPlatform.system}" or (throw "Unsupported system for ${pname}: ${stdenv.hostPlatform.system}");
```

File: pkgs/development/python-modules/kaleido/default.nix [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+  nativeBuildInputs = [ autoPatchelfHook ];
```
Comment (natsukium) [inline pkgs/development/python-modules/kaleido/default.nix:44]:
```suggestion
  nativeBuildInputs = lib.optionals stdenv.hostPlatform.isLinux [ autoPatchelfHook ];
```

File: pkgs/development/python-modules/kaleido/default.nix [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    description = "Fast static image export for web-based visualization libraries with zero dependencies";
+    homepage = "https://github.com/plotly/Kaleido";
+    changelog = "https://github.com/plotly/Kaleido/releases";
+    platforms = [
+      "x86_64-linux"
+      "x86_64-darwin"
+      "aarch64-linux"
+      "aaarch64-darwin"
+    ];
+    sourceProvenance = with lib.sourceTypes; [ binaryNativeCode ]; # Trust me, I'm not happy. But after literal hours of trying to reverse-engineer their build system and getting nowhere, I'll use the stupid binaries >:(
```
Comment (natsukium) [inline pkgs/development/python-modules/kaleido/default.nix:73]: This is the third attempt, as far as I know, but no one has been able to build it from the source.😢

File: pkgs/development/python-modules/kaleido/default.nix [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+  lib,
+  buildPythonPackage,
+  fetchurl,
+  autoPatchelfHook,
+  expat,
+  libGL,
+  nspr,
+  nss,
+  sqlite,
+  system ? builtins.currentSystem,
```
Comment (eclairevoyant) [inline pkgs/development/python-modules/kaleido/default.nix:11]:
```suggestion
```

File: pkgs/development/python-modules/kaleido/default.nix [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    in
+    {
+      # This library is so cursed that I have to use fetchurl instead of fetchPypi. I am not happy.
+      x86_64-linux = fetchurl {
+        url = "https://files.pythonhosted.org/packages/py2.py3/k/kaleido/kaleido-0.2.1-py2.py3-none-manylinux1_x86_64.whl";
+        hash = "sha256-qiHPG/HHj4+lCp99ReEAPDh709b+CnZ8+780S5W9w6g=";
+      };
+      aarch64-linux = fetchurl {
+        url = "https://files.pythonhosted.org/packages/py2.py3/k/kaleido/kaleido-0.2.1.post1-py2.py3-none-manylinux2014_armv7l.whl";
+        hash = lib.fakeHash;
```
Comment (eclairevoyant) [inline pkgs/development/python-modules/kaleido/default.nix:31]: These should be replaced

File: pkgs/development/python-modules/kaleido/default.nix (hunk 1 above)
Comment (Pandapip1) [reply pkgs/development/python-modules/kaleido/default.nix:45]: I haven't used `passthru.tests` at all yet. Got any pointers?

Comment (eclairevoyant) [reply pkgs/development/python-modules/kaleido/default.nix:43]: `nix-build -A python3Packages.kaleido.tests`

Comment (Pandapip1) [reply pkgs/development/python-modules/kaleido/default.nix:43]: Thanks! I'll look into it.

Comment (Pandapip1) [reply pkgs/development/python-modules/kaleido/default.nix:45]: Done!

**Comment by eclairevoyant** [comment]
Also please use the full attrname in commits so that borgo knows what to build (i.e. `python312Packages.kaleido` and same with plotly)

**Comment by Pandapip1** [comment]
Just noticed there weren't any maintainers for plotly, so this PR also has me adopting it.

**Comment by eclairevoyant** [comment]
please use python312Packages, not python3Packages.

**Comment by natsukium** [comment]
@ofborg build python312Packages.kaleido python311Packages.kaleido

**Review by natsukium — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/kaleido/default.nix [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+  meta = {
+    description = "Fast static image export for web-based visualization libraries with zero dependencies";
+    homepage = "https://github.com/plotly/Kaleido";
+    changelog = "https://github.com/plotly/Kaleido/releases";
+    platforms = [
+      "x86_64-linux"
+      "x86_64-darwin"
+      "aarch64-linux"
+      "aaarch64-darwin"
```
[the remark this answers is not in the document]
Comment (Pandapip1) [reply pkgs/development/python-modules/kaleido/default.nix:74]: Oops!

**Comment by eclairevoyant** [comment]
@Pandapip1 please fix the commit messages.

**Comment by Pandapip1** [comment]
@eclairevoyant aren't they fixed? They're in the `python3Packages.xxx` format.

**Comment by Pandapip1** [comment]
I think I figured out what you meant. Does ofborg not recognize the `python3Packages.xxx` format?

File: pkgs/development/python-modules/kaleido/tests.nix [hunk 11]
```diff
@@ -0,0 +1,15 @@
+{
+  runCommand,
+  python3,
```
Comment (eclairevoyant) [inline pkgs/development/python-modules/kaleido/tests.nix:3]:
```suggestion
  python,
```

Missed this earlier, oops. (We really should have a check for this.)

File: pkgs/development/python-modules/kaleido/tests.nix [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+  plotly,
+  pandas,
+  kaleido,
+}:
+
+runCommand "${kaleido.pname}-tests" {
+  nativeBuildInputs = [
+    python3
```
Comment (eclairevoyant) [inline pkgs/development/python-modules/kaleido/tests.nix:11]:
```suggestion
    python
```

**Comment by eclairevoyant** [comment]
> I think I figured out what you meant. Does ofborg not recognize the `python3Packages.xxx` format?

Right, hence the note here to explicitly specify the version:
https://github.com/NixOS/nixpkgs/blob/105030f1b542eb3b353f027f53c2dc4a0d1f4618/doc/languages-frameworks/python.section.md?plain=1#L2020-L2023

**Comment by Pandapip1** [comment]
Okay, that's not particularly clear. I'll open a PR to rephrase that.

**Comment by natsukium** [comment]
Tests fail on darwin with the following log.
 `*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Critical error: required built-in appearance SystemAppearance not found'`

I suspect `CoreFoundation`, `AppKit`, or the matplotlib environment variable `MPLBACKEND`, but I can't fix it.
Maybe someone at @NixOS/darwin-maintainers  knows how to make it work.

<summary></summary><details>

```
kaleido-tests-aarch64-darwin> building '/nix/store/r6dib0z8rc7pbjhfqrh184a6vh9k9a8h-kaleido-tests.drv'
kaleido-tests-aarch64-darwin> Traceback (most recent call last):
kaleido-tests-aarch64-darwin>   File "/nix/store/i36nyn014clgrvrgmv60vg58wl0ifmvn-tests.py", line 11, in <module>
kaleido-tests-aarch64-darwin>     fig.write_image(outfile, engine="kaleido")
kaleido-tests-aarch64-darwin>   File "/nix/store/k4hlqc8dcdw3890fvghzl3sppwfcs78p-python3.12-plotly-5.23.0/lib/python3.12/site-packages/plotly/basedatatypes.py", line 3835, in write_image
kaleido-tests-aarch64-darwin>     return pio.write_image(self, *args, **kwargs)
kaleido-tests-aarch64-darwin>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
kaleido-tests-aarch64-darwin>   File "/nix/store/k4hlqc8dcdw3890fvghzl3sppwfcs78p-python3.12-plotly-5.23.0/lib/python3.12/site-packages/plotly/io/_kaleido.py", line 266, in write_image
kaleido-tests-aarch64-darwin>     img_data = to_image(
kaleido-tests-aarch64-darwin>                ^^^^^^^^^
kaleido-tests-aarch64-darwin>   File "/nix/store/k4hlqc8dcdw3890fvghzl3sppwfcs78p-python3.12-plotly-5.23.0/lib/python3.12/site-packages/plotly/io/_kaleido.py", line 143, in to_image
kaleido-tests-aarch64-darwin>     img_bytes = scope.transform(
kaleido-tests-aarch64-darwin>                 ^^^^^^^^^^^^^^^^
kaleido-tests-aarch64-darwin>   File "/nix/store/y1fsgdcc12c2jyb1pal6ds0icrz7x8x2-python3.12-kaleido-0.2.1/lib/python3.12/site-packages/kaleido/scopes/plotly.py", line 153, in transform
kaleido-tests-aarch64-darwin>     response = self._perform_transform(
kaleido-tests-aarch64-darwin>                ^^^^^^^^^^^^^^^^^^^^^^^^
kaleido-tests-aarch64-darwin>   File "/nix/store/y1fsgdcc12c2jyb1pal6ds0icrz7x8x2-python3.12-kaleido-0.2.1/lib/python3.12/site-packages/kaleido/scopes/base.py", line 293, in _perform_transform
kaleido-tests-aarch64-darwin>     self._ensure_kaleido()
kaleido-tests-aarch64-darwin>   File "/nix/store/y1fsgdcc12c2jyb1pal6ds0icrz7x8x2-python3.12-kaleido-0.2.1/lib/python3.12/site-packages/kaleido/scopes/base.py", line 198, in _ensure_kaleido
kaleido-tests-aarch64-darwin>     raise ValueError(message)
kaleido-tests-aarch64-darwin> ValueError: Failed to start Kaleido subprocess. Error stream:
kaleido-tests-aarch64-darwin> 
kaleido-tests-aarch64-darwin> [0905/000022.334097:WARNING:resource_bundle.cc(431)] locale_file_path.empty() for locale en-US
kaleido-tests-aarch64-darwin> *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Critical error: required built-in appearance SystemAppearance not found'
kaleido-tests-aarch64-darwin> *** First throw call stack:
kaleido-tests-aarch64-darwin> (
kaleido-tests-aarch64-darwin> 	0   CoreFoundation                      0x0000000186e662ec __exceptionPreprocess + 176
kaleido-tests-aarch64-darwin> 	1   libobjc.A.dylib                     0x000000018694a158 objc_exception_throw + 60
kaleido-tests-aarch64-darwin> 	2   CoreFoundation                      0x0000000186e661dc +[NSException exceptionWithName:reason:userInfo:] + 0
kaleido-tests-aarch64-darwin> 	3   AppKit                              0x000000018a61e214 __33+[NSAppearance _initializeCoreUI]_block_invoke + 88
kaleido-tests-aarch64-darwin> 	4   libdispatch.dylib                   0x0000000186b613e8 _dispatch_client_callout + 20
kaleido-tests-aarch64-darwin> 	5   libdispatch.dylib                   0x0000000186b62c68 _dispatch_once_callout + 32
kaleido-tests-aarch64-darwin> 	6   AppKit                              0x000000018a63e5cc +[NSAppearance _aquaAppearance] + 60
kaleido-tests-aarch64-darwin> 	7   AppKit                              0x000000018a61d98c +[NSAppearance appearanceNamed:] + 32
kaleido-tests-aarch64-darwin> 	8   AppKit                              0x000000018a61d124 -[NSSystemAppearanceProxy init] + 124
kaleido-tests-aarch64-darwin> 	9   AppKit                              0x000000018a61d098 __38+[NSSystemAppearanceProxy systemProxy]_block_invoke + 24
kaleido-tests-aarch64-darwin> 	10  libdispatch.dylib                   0x0000000186b613e8 _dispatch_client_callout + 20
kaleido-tests-aarch64-darwin> 	11  libdispatch.dylib                   0x0000000186b62c68 _dispatch_once_callout + 32
kaleido-tests-aarch64-darwin> 	12  AppKit                              0x000000018a61d07c +[NSSystemAppearanceProxy systemProxy] + 64
kaleido-tests-aarch64-darwin> 	13  AppKit                              0x000000018a61d008 -[NSApplication(NSApplicationAppearance_Internal) _registerForAppearanceNotifications] + 32
kaleido-tests-aarch64-darwin> 	14  AppKit                              0x000000018a61acec -[NSApplication init] + 904
kaleido-tests-aarch64-darwin> 	15  AppKit                              0x000000018a61a798 +[NSApplication sharedApplication] + 128
kaleido-tests-aarch64-darwin> 	16  kaleido                             0x000000010908fb74 _ZN8headless27HeadlessContentMainDelegate24PreCreateMainMessageLoopEv + 36
kaleido-tests-aarch64-darwin> 	17  kaleido                             0x0000000106fb9748 _ZN7content21ContentMainRunnerImpl17RunServiceManagerERNS_18MainFunctionParamsEb + 188
kaleido-tests-aarch64-darwin> 	18  kaleido                             0x0000000106fb9660 _ZN7content21ContentMainRunnerImpl3RunEb + 304
kaleido-tests-aarch64-darwin> 	19  kaleido                             0x0000000106fb7e40 _ZN7content17RunContentProcessERKNS_17ContentMainParamsEPNS_17ContentMainRunnerE + 552
kaleido-tests-aarch64-darwin> 	20  kaleido                             0x0000000106fb8520 _ZN7content11ContentMainERKNS_17ContentMainParamsE + 48
kaleido-tests-aarch64-darwin> 	21  kaleido                             0x0000000107041240 _ZN8headless12_GLOBAL__N_114RunContentMainENS_15HeadlessBrowser7OptionsEN4base12OnceCallbackIFvPS1_EEE + 260
kaleido-tests-aarch64-darwin> 	22  kaleido                             0x0000000107041104 _ZN8headless19HeadlessBrowserMainENS_15HeadlessBrowser7OptionsEN4base12OnceCallbackIFvPS0_EEE + 56
kaleido-tests-aarch64-darwin> 	23  kaleido                             0x000000010444b8e4 main + 180
kaleido-tests-aarch64-darwin> 	24  dyld                                0x0000000186987154 start + 2476
kaleido-tests-aarch64-darwin> )
kaleido-tests-aarch64-darwin> /nix/store/y1fsgdcc12c2jyb1pal6ds0icrz7x8x2-python3.12-kaleido-0.2.1/lib/python3.12/site-packages/kaleido/executable/kaleido: line 5:  3017 Trace/BPT trap: 5       ./bin/kaleido $@
```
</details>

**Comment by Pandapip1** [comment]
@natsukium in the meantime, I've marked the package as broken on darwin. Does that address that concern for the time being?

**Comment by natsukium** [comment]
I think it looks good. The darwin fix can be done in another PR.

**Comment by natsukium** [comment]
cc @GaetanLepage, @Sigmanificient if you're interested.

**Review by natsukium — APPROVED** [review]

**Comment by Pandapip1** [comment]
Result of `nixpkgs-review pr 339136` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages marked as broken and skipped:</summary>
  <ul>
    <li>picosnitch</li>
    <li>picosnitch.dist</li>
    <li>python311Packages.pymatgen</li>
    <li>python311Packages.pymatgen.dist</li>
    <li>python311Packages.sumo</li>
    <li>python311Packages.sumo.dist</li>
    <li>python312Packages.pymatgen</li>
    <li>python312Packages.pymatgen.dist</li>
    <li>python312Packages.sumo</li>
    <li>python312Packages.sumo.dist</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>theharvester</li>
    <li>theharvester.dist</li>
  </ul>
</details>
<details>
  <summary>120 packages built:</summary>
  <ul>
    <li>apkleaks</li>
    <li>apkleaks.dist</li>
    <li>checkov</li>
    <li>checkov.dist</li>
    <li>cve-bin-tool</li>
    <li>cve-bin-tool.dist</li>
    <li>das</li>
    <li>das.dist</li>
    <li>deeptools</li>
    <li>deeptools.dist</li>
    <li>hp2p</li>
    <li>jadx</li>
    <li>python311Packages.ax-platform</li>
    <li>python311Packages.ax-platform.dist</li>
    <li>python311Packages.catboost</li>
    <li>python311Packages.catboost.dist</li>
    <li>python311Packages.chart-studio</li>
    <li>python311Packages.chart-studio.dist</li>
    <li>python311Packages.compressai</li>
    <li>python311Packages.compressai.dist</li>
    <li>python311Packages.cufflinks</li>
    <li>python311Packages.cufflinks.dist</li>
    <li>python311Packages.dash</li>
    <li>python311Packages.dash-bootstrap-components</li>
    <li>python311Packages.dash-bootstrap-components.dist</li>
    <li>python311Packages.dash.dist</li>
    <li>python311Packages.experiment-utilities</li>
    <li>python311Packages.experiment-utilities.dist</li>
    <li>python311Packages.explorerscript</li>
    <li>python311Packages.explorerscript.dist</li>
    <li>python311Packages.igraph</li>
    <li>python311Packages.igraph.dist</li>
    <li>python311Packages.kaleido</li>
    <li>python311Packages.kaleido.dist</li>
    <li>python311Packages.kmapper</li>
    <li>python311Packages.kmapper.dist</li>
    <li>python311Packages.leidenalg</li>
    <li>python311Packages.leidenalg.dist</li>
    <li>python311Packages.meshtastic</li>
    <li>python311Packages.meshtastic.dist</li>
    <li>python311Packages.optuna</li>
    <li>python311Packages.optuna.dist</li>
    <li>python311Packages.orange3</li>
    <li>python311Packages.orange3.dist</li>
    <li>python311Packages.plotly</li>
    <li>python311Packages.plotly.dist</li>
    <li>python311Packages.pyannote-audio</li>
    <li>python311Packages.pyannote-audio.dist</li>
    <li>python311Packages.pyannote-pipeline</li>
    <li>python311Packages.pyannote-pipeline.dist</li>
    <li>python311Packages.reflex</li>
    <li>python311Packages.reflex.dist</li>
    <li>python311Packages.scikit-tda</li>
    <li>python311Packages.scikit-tda.dist</li>
    <li>python311Packages.skytemple-dtef</li>
    <li>python311Packages.skytemple-dtef.dist</li>
    <li>python311Packages.skytemple-files</li>
    <li>python311Packages.skytemple-files.dist</li>
    <li>python311Packages.skytemple-ssb-debugger</li>
    <li>python311Packages.skytemple-ssb-debugger.dist</li>
    <li>python311Packages.synergy</li>
    <li>python311Packages.synergy.dist</li>
    <li>python311Packages.textnets</li>
    <li>python311Packages.textnets.dist</li>
    <li>python312Packages.ax-platform</li>
    <li>python312Packages.ax-platform.dist</li>
    <li>python312Packages.catboost</li>
    <li>python312Packages.catboost.dist</li>
    <li>python312Packages.chart-studio</li>
    <li>python312Packages.chart-studio.dist</li>
    <li>python312Packages.compressai</li>
    <li>python312Packages.compressai.dist</li>
    <li>python312Packages.cufflinks</li>
    <li>python312Packages.cufflinks.dist</li>
    <li>python312Packages.dash</li>
    <li>python312Packages.dash-bootstrap-components</li>
    <li>python312Packages.dash-bootstrap-components.dist</li>
    <li>python312Packages.dash.dist</li>
    <li>python312Packages.experiment-utilities</li>
    <li>python312Packages.experiment-utilities.dist</li>
    <li>python312Packages.explorerscript</li>
    <li>python312Packages.explorerscript.dist</li>
    <li>python312Packages.igraph</li>
    <li>python312Packages.igraph.dist</li>
    <li>python312Packages.kaleido</li>
    <li>python312Packages.kaleido.dist</li>
    <li>python312Packages.kmapper</li>
    <li>python312Packages.kmapper.dist</li>
    <li>python312Packages.leidenalg</li>
    <li>python312Packages.leidenalg.dist</li>
    <li>python312Packages.meshtastic</li>
    <li>python312Packages.meshtastic.dist</li>
    <li>python312Packages.optuna</li>
    <li>python312Packages.optuna.dist</li>
    <li>python312Packages.orange3</li>
    <li>python312Packages.orange3.dist</li>
    <li>python312Packages.plotly</li>
    <li>python312Packages.plotly.dist</li>
    <li>python312Packages.pyannote-audio</li>
    <li>python312Packages.pyannote-audio.dist</li>
    <li>python312Packages.pyannote-pipeline</li>
    <li>python312Packages.pyannote-pipeline.dist</li>
    <li>python312Packages.reflex</li>
    <li>python312Packages.reflex.dist</li>
    <li>python312Packages.scikit-tda</li>
    <li>python312Packages.scikit-tda.dist</li>
    <li>python312Packages.skytemple-dtef</li>
    <li>python312Packages.skytemple-dtef.dist</li>
    <li>python312Packages.skytemple-files</li>
    <li>python312Packages.skytemple-files.dist</li>
    <li>python312Packages.skytemple-ssb-debugger</li>
    <li>python312Packages.skytemple-ssb-debugger.dist</li>
    <li>python312Packages.synergy</li>
    <li>python312Packages.synergy.dist</li>
    <li>quark-engine</li>
    <li>quark-engine.dist</li>
    <li>skytemple</li>
    <li>skytemple.dist</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
  </ul>
</details>

**Comment by Pandapip1** [comment]
Manually backported: https://github.com/NixOS/nixpkgs/pull/339839

**Comment by sarahec** [comment]
Here's a weirdness: it builds fine on Darwin when run as a stand-alone build. The error you saw (essentially "we don't know how to find the system UI settings") may be due to the building machine being headless.

Recommendation: disable the failing test. It passes when run on a desktop/laptop.

**Comment by emilazy** [comment]
It’s just the Nix sandbox forbidding access to OS stuff. Happens a lot.

**Comment by Pandapip1** [comment]
Unfortunately I don't have access to a darwin machine to debug this. I suspect it might be `libGL`-related?

**Comment by sarahec** [comment]
> Unfortunately I don't have access to a darwin machine to debug this. I suspect it might be `libGL`-related?

I think @emilazy has identified the issue. I'm attempting to verify it on a Darwin machine so we can disable the appropriate tests.

**Comment by Pandapip1** [comment]
> I think @emilazy has identified the issue.

Yea, I was just noting that `libGL` is one of the few impurities left in nixpkgs.

**Comment by emilazy** [comment]
I’m not sure what you mean; `libGL` isn’t impure. On Darwin, it’s the same as `darwin.apple_sdk.frameworks.OpenGL`.

**Comment by eclairevoyant** [comment]
Loading libGL is done impurely in NixOS, because nvidia.
(Not sure if the error is related since I can't find it atm.)

**Comment by emilazy** [comment]
I wouldn’t quite call that an impurity. `libGL` itself is `libglvnd`, a perfectly pure, standard, Nix‐packaged library. That library happens to delegate to a system OpenGL driver at runtime, which I would say makes it *late‐bound*, rather than impure per se; the properties of the Nix build aren’t being violated at any point.

Anyway, that’s not too relevant to macOS, where every system framework is built inside Nix against purely‐packaged stubs listing the available symbols, and which is then linked against the OS‐level framework code at runtime by the dynamic loader. (So in a sense, every system library is like `libGL` on Darwin.)

**Comment by Pandapip1** [comment]
> I wouldn’t quite call that an impurity. libGL itself is libglvnd, a perfectly pure, standard, Nix‐packaged library. That library happens to delegate to a system OpenGL driver at runtime, which I would say makes it late‐bound, rather than impure per se; the properties of the Nix build aren’t being violated at any point.

`libglvnd`'s `.so`s have `RPATH`s that contain paths outside of `/nix/store`.

**Comment by sarahec** [comment]
> I think @emilazy has identified the issue. I'm attempting to verify it on a Darwin machine so we can disable the appropriate tests.

I can't get the build to break for me on Darwin, even with sandboxing enabled. Should I file a PR that conditionally disables the passthrough test on Darwin and removes the `broken` flag?

**Comment by Pandapip1** [comment]
> I can't get the build to break for me on Darwin, even with sandboxing enabled. Should I file a PR that conditionally disables the passthrough test on Darwin and removes the broken flag?

Go right ahead!

**Pull request merged**